File: java/Easy/_876_MiddleoftheLinkedList.java
package Easy;

import org.testng.annotations.Test;

import CS.ListNode;

public class _876_MiddleoftheLinkedList {
    public ListNode middleNode(ListNode head) {
        ListNode cur1 = head;
        ListNode cur2 = head;

        while(cur1.next != null){
            if(cur2 == null || cur2.next == null) return cur1;
            cur1 = cur1.next;
            cur2 = cur2.next.next;
        }
        return cur1;
    }

    @Test
    public void test(){
        ListNode test = new ListNode(1);
        test.next = new ListNode(2);
        test.next.next = new ListNode(3);
        test.next.next.next = new ListNode(4);
        test.next.next.next.next = new ListNode(5);
        middleNode(test);
    }
}

File: Python3/_876_MiddleOfTheLinkedList.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next
class Solution:
    def middleNode(self, head: Optional[ListNode]) -> Optional[ListNode]:
        cur1 = head
        cur2 = head
        while cur2 != None and cur2.next != None:
            cur1 = cur1.next
            cur2 = cur2.next.next
        return cur1